off_chain_data/application-go/blockParser.go

```diff
@@ -0,0 +1,397 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/hyperledger/fabric-gateway/pkg/identity"
+	"github.com/hyperledger/fabric-protos-go-apiv2/common"
+	"github.com/hyperledger/fabric-protos-go-apiv2/ledger/rwset"
+	"github.com/hyperledger/fabric-protos-go-apiv2/ledger/rwset/kvrwset"
+	"github.com/hyperledger/fabric-protos-go-apiv2/msp"
+	"github.com/hyperledger/fabric-protos-go-apiv2/peer"
+	"google.golang.org/protobuf/proto"
+)
+
+type Block interface {
+	GetNumber() uint64
+	GetTransactions() []Transaction
+	ToProto() *common.Block
+}
+
+type ParsedBlock struct {
+	block           *common.Block
+	validationCodes []byte
+	transactions    []Transaction
+}
+
+func NewParsedBlock(block *common.Block) Block {
+	validationCodes := getTransactionValidationCodes(block)
+
+	return &ParsedBlock{block, validationCodes, nil}
+}
```

Comment:
A more typical pattern in Go is to return concrete structs, and use interfaces for parameter types received by functions. Interfaces are typically used for return types only where different implementation structs might be returned.

For this case you might be better not to have an interface at all.

```suggestion
type Block struct {
	block           *common.Block
	validationCodes []byte
	transactions    []Transaction
}

func ParseBlock(block *common.Block) Block {
	validationCodes := getTransactionValidationCodes(block)

	return &Block{block, validationCodes, nil}
}
```

To make it clearer you might consider moving the parsing code (and associated structs) into a different package. Then each Block variable in the application code will be referred to as either `common.Block` or `parser.Block` (for example).

Reply:
> A more typical pattern in Go is to return concrete structs, and use interfaces for parameter types received by functions.

Thx, didn't knew. Will do that.

> Interfaces are typically used for return types only where different implementation structs might be returned.

Of course, that is right and not only in go. Will keep it in mind.

> To make it clearer you might consider moving the parsing code (and associated structs) into a different package. Then each Block variable in the application code will be referred to as either `common.Block` or `parser.Block` (for example).

Yes, that is exactly my plan once I make it work.

Reply:
> For this case you might be better not to have an interface at all.

The TypeScript implementation is using interfaces. I thought I was supposed to follow it closely?

I will implement a version without interfaces and see how it is.